price-estimator/src/amounts_at_price.rs

```diff
@@ -0,0 +1,116 @@
+use pricegraph::{Pricegraph, TokenPair, TransitiveOrder};
+
+/// An overlapping order where buy / sell == price and the amounts take into account that the solver
+/// will subtract the rounding buffer from the sell amount.
+pub fn order_at_price_with_rounding_buffer(
+    token_pair: TokenPair,
+    limit_price: f64,
+    pricegraph: &Pricegraph,
+    rounding_buffer: f64,
+) -> Option<TransitiveOrder> {
+    let order = pricegraph.order_for_limit_price(token_pair, limit_price)?;
+    // We know that an order is still overlapping if it has a limit price <= this one. The limit
+    // price of this order is usually larger (better for the seller) than the user requested limit
+    // price.
+    let order_limit_price = order.buy / order.sell;
+
+    // The order that the user places uses the requested limit price exactly. It is less restrictive
+    // (still overlapping) than the pricegraph order because the buy amount is lower.
+    let order_that_user_places = TransitiveOrder {
+        sell: order.sell,
+        buy: order.sell * limit_price,
+    };
+
+    // The solver sees the user order with a slightly lower sell amount.
+    let limit_price_that_solver_sees =
+        order_that_user_places.buy / (order_that_user_places.sell - rounding_buffer);
+    if limit_price_that_solver_sees <= order_limit_price {
+        // If that price is not more restrictive than the pricegraph order's limit price then it is
+        // still overlapping and we can use it. This case it much more likely than the other branch
+        // because you would have to hit a narrow price so that there is no room to adjust for the
+        // small (compared to buy and sell amounts) rounding buffer.
+        Some(order_that_user_places)
+    } else {
+        // Otherwise we cannot keep the limit_price exactly the same as requested. Since rounding
+        // buffers are small compared to traded amounts the introduced error is small.
+        Some(TransitiveOrder {
+            sell: order_that_user_places.sell + rounding_buffer,
+            buy: order_that_user_places.buy,
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use pricegraph::{Element, PriceFraction, TokenPair, UserId, Validity, FEE_FACTOR};
+
+    #[test]
+    fn order_at_price_likely_branch() {
```

Comment:
This *is* testing an implementation detail. The comment might not stay up to date but likewise the function body could change so that the distinction isn't relevant. Anyway, I'll change the name.

Reply:
Thanks. Generally I think it's nice if the test name can allude to the expectation that is being tested (e.g. what is the given & expected behavior) since it's not always obvious from looking at the setup & asserts. At least for me it helps understanding the intended behaviour when trying to fix failing tests after a refactor.